// @data-client/core/legacy/manager/DefaultConnectionListener.js
export class BrowserConnectionListener {
  isOnline() {
    if (navigator.onLine !== undefined) {
      return navigator.onLine;
    }
    return true;
  }
  addOnlineListener(handler) {
    addEventListener('online', handler);
  }
  removeOnlineListener(handler) {
    removeEventListener('online', handler);
  }
  addOfflineListener(handler) {
    addEventListener('offline', handler);
  }
  removeOfflineListener(handler) {
    removeEventListener('offline', handler);
  }
}
export class AlwaysOnlineConnectionListener {
  isOnline() {
    /* istanbul ignore next */
    return true;
  }
  addOnlineListener() {}
  removeOnlineListener() {}
  addOfflineListener() {}
  removeOfflineListener() {}
}
let DefaultConnectionListener;
/* istanbul ignore if */
if (typeof navigator !== 'undefined' && typeof addEventListener === 'function') {
  DefaultConnectionListener = BrowserConnectionListener;
} else {
  /* istanbul ignore next */
  DefaultConnectionListener = AlwaysOnlineConnectionListener;
}
export default DefaultConnectionListener;
