// @data-client/core/legacy/manager/DevtoolsManager.js
var _globalThis$document;
import createReducer from '../state/reducer/createReducer.js';
const HASINTL = typeof Intl !== 'undefined';
const DEFAULT_CONFIG = {
  name: `Rest Hooks: ${(_globalThis$document = globalThis.document) == null ? void 0 : _globalThis$document.title}`,
  autoPause: true,
  serialize: {
    options: undefined,
    /* istanbul ignore next */
    replacer: HASINTL ? (key, value) => {
      if (typeof value === 'number' && typeof key === 'string' && isFinite(value) && (key === 'date' || key.endsWith('At'))) {
        return Intl.DateTimeFormat('en-US', {
          hour: 'numeric',
          minute: 'numeric',
          second: 'numeric',
          fractionalSecondDigits: 3
        }).format(value);
      }
      return value;
    } : undefined
  }
};

/** Integrates with https://github.com/reduxjs/redux-devtools
 *
 * Options: https://github.com/reduxjs/redux-devtools/blob/main/extension/docs/API/Arguments.md
 *
 * @see https://dataclient.io/docs/api/DevToolsManager
 */
export default class DevToolsManager {
  constructor(config, skipLogging) {
    /* istanbul ignore next */
    this.devTools = typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__.connect(Object.assign({}, DEFAULT_CONFIG, {
      config
    }));

    /* istanbul ignore if */
    /* istanbul ignore next */
    if (this.devTools) {
      this.middleware = controller => {
        const reducer = createReducer(controller);
        return next => action => {
          const ret = next(action);
          ret.then(() => {
            if (skipLogging != null && skipLogging(action)) return;
            const state = controller.getState();
            this.devTools.send(action, state.optimistic.reduce(reducer, state), undefined, 'REST_HOOKS');
          });
          return ret;
        };
      };
    } else {
      this.middleware = () => next => action => next(action);
    }
  }

  /** Called when initial state is ready */
  init(state) {
    /* istanbul ignore if */
    if (this.devTools) this.devTools.init(state);
  }

  /** Ensures all subscriptions are cleaned up. */
  cleanup() {}

  /** Attaches Manager to store
   *
   */
  getMiddleware() {
    return this.middleware;
  }
}
