// @data-client/core/lib/state/reducer/createReducer.js
import { expireReducer } from './expireReducer.js';
import { fetchReducer } from './fetchReducer.js';
import { invalidateReducer } from './invalidateReducer.js';
import { setReducer } from './setReducer.js';
import { SET_TYPE, INVALIDATE_TYPE, RESET_TYPE, FETCH_TYPE, GC_TYPE, OPTIMISTIC_TYPE, INVALIDATEALL_TYPE, EXPIREALL_TYPE } from '../../actionTypes.js';
export default function createReducer(controller) {
  return function reducer(state, action) {
    if (!state) state = initialState;
    switch (action.type) {
      case GC_TYPE:
        // inline deletes are fine as these should have 0 refcounts
        action.entities.forEach(([key, pk]) => {
          var _entities$key, _entityMeta$key;
          (_entities$key = state.entities[key]) == null ? true : delete _entities$key[pk];
          (_entityMeta$key = state.entityMeta[key]) == null ? true : delete _entityMeta$key[pk];
        });
        action.results.forEach(fetchKey => {
          delete state.results[fetchKey];
          delete state.meta[fetchKey];
        });
        return state;
      case FETCH_TYPE:
        return fetchReducer(state, action);
      case OPTIMISTIC_TYPE:
      // eslint-disable-next-line no-fallthrough
      case SET_TYPE:
        return setReducer(state, action, controller);
      case INVALIDATEALL_TYPE:
      case INVALIDATE_TYPE:
        return invalidateReducer(state, action);
      case EXPIREALL_TYPE:
        return expireReducer(state, action);
      case RESET_TYPE:
        return {
          ...initialState,
          lastReset: action.date
        };
      default:
        // A reducer must always return a valid state.
        // Alternatively you can throw an error if an invalid action is dispatched.
        return state;
    }
  };
}
export const initialState = {
  entities: {},
  indexes: {},
  results: {},
  meta: {},
  entityMeta: {},
  optimistic: [],
  lastReset: 0
};
