// @data-client/normalizr/lib/WeakEntityMap.js
import { isImmutable } from './schemas/ImmutableUtils.js';
/** Maps entity dependencies to a value (usually their denormalized form)
 *
 * Dependencies store `Path` to enable quick traversal using only `State`
 * If *any* members of the dependency get cleaned up, so does that key/value pair get removed.
 */
export default class WeakEntityMap {
  constructor() {
    this.next = new WeakMap();
    this.nextPath = undefined;
  }
  get(entity, getEntity) {
    let curLink = this.next.get(entity);
    if (!curLink) return EMPTY;
    while (curLink.nextPath) {
      const nextEntity = getEntity(curLink.nextPath);
      curLink = curLink.next.get(nextEntity);
      if (!curLink) return EMPTY;
    }
    // curLink exists, but has no path - so must have a value
    return [curLink.value, curLink.journey];
  }
  set(dependencies, value) {
    if (dependencies.length < 1) throw new KeySize();
    let curLink = this;
    for (const {
      entity,
      path
    } of dependencies) {
      let nextLink = curLink.next.get(entity);
      if (!nextLink) {
        nextLink = new Link();
        curLink.next.set(entity, nextLink);
      }
      curLink.nextPath = path;
      curLink = nextLink;
    }
    // in case there used to be more
    curLink.nextPath = undefined;
    curLink.value = value;
    // we could recompute this on get, but it would have a cost and we optimize for `get`
    curLink.journey = depToPaths(dependencies);
  }
}
const EMPTY = [undefined, undefined];
export function getEntities(state) {
  const entityIsImmutable = isImmutable(state);
  if (entityIsImmutable) {
    return ({
      key,
      pk
    }) => state.getIn([key, pk]);
  } else {
    return ({
      key,
      pk
    }) => {
      var _state$key;
      return (_state$key = state[key]) == null ? void 0 : _state$key[pk];
    };
  }
}
export function depToPaths(dependencies) {
  var _paths$;
  const paths = dependencies.map(dep => dep.path);
  // the first item is the result input not an actual entity
  if (((_paths$ = paths[0]) == null ? void 0 : _paths$.pk) === '') paths.shift();
  return paths;
}
/** Link in a chain */
class Link {
  constructor() {
    this.next = new WeakMap();
    this.value = undefined;
    this.journey = [];
    this.nextPath = undefined;
  }
}
class KeySize extends Error {
  constructor(...args) {
    super(...args);
    this.message = 'Keys must include at least one member';
  }
}
